src/main/java/io/fabric8/elasticsearch/plugin/acl/ACLDocumentManager.java

```diff
@@ -0,0 +1,283 @@
+                    SearchGuardRolesMapping mappings = (SearchGuardRolesMapping) doc;
+                    for (RolesMapping mapping : mappings) {
+                        //assume if the value is there its intentional
+                        if(mapping.getExpire() != null && mapping.getExpire().longValue() < now) {
+                            mappings.removeRolesMapping(mapping);
+                        }
+                    }
+                } else if(ConfigurationSettings.SEARCHGUARD_ROLE_TYPE.equals(doc.getType())) {
+                    SearchGuardRoles roles = (SearchGuardRoles) doc;
+                    for (Roles role : roles) {
+                        //assume if the value is there its intentional
+                        if(role.getExpire() != null && role.getExpire().longValue() < now) {
+                            roles.removeRole(role);
+                        }
+                    }
+                }
+            }
+        }
+
+        @Override
+        public BulkRequest buildRequest(Client client, BulkRequestBuilder builder, Collection<SearchGuardACLDocument> docs) throws IOException{
+            for (SearchGuardACLDocument doc : docs) {
+                IndexRequestBuilder indexBuilder = client
+                        .prepareIndex(searchGuardIndex, doc.getType(), SEARCHGUARD_CONFIG_ID)
+                        .setConsistencyLevel(WriteConsistencyLevel.DEFAULT)
+                        .setOpType(OpType.INDEX)
+                        .setSource(doc.toXContentBuilder());
+                builder.add(indexBuilder.request());
+                if(LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Built {} index request: {}", doc.getType(), XContentHelper.convertToJson(doc.toXContentBuilder().bytes(),true, true));
+                }
+            }
+            return builder.request();
+        }
+        
+        
+        
+    }
+
+    public void expire() {
+        syncAcl(new ExpireOperation(System.currentTimeMillis()));
+    }
+
+    public void syncAcl(OpenshiftRequestContext context) {
+        if(!syncAcl(new SyncFromContextOperation(context))){
+            LOGGER.warn("Unable to sync ACLs for request from user: {}", context.getUser());
+        }
+    }    
+    
+    private boolean syncAcl(ACLDocumentOperation operation) {
+        //try up to 30 seconds and then continue
+        for (int n : new int [] {1 , 1 , 2 , 3 , 5 , 8}) {
+            if(trySyncAcl(operation)) {
+                return true;
+            }
+            try {
+                if(LOGGER.isTraceEnabled()) {
+                    LOGGER.trace("Sleeping for {}(s)", n * 1000);
+                }
+                Thread.sleep(n * 1000);
+            } catch (InterruptedException e) {
+                LOGGER.error("There was an error while trying the sleep the syncACL operation", e);
+            }
+        }
+        return false;
+    }
+    
+    public boolean trySyncAcl(ACLDocumentOperation operation) {
```

Comment:
could we make this method sychronized instead of using locks?

Reply:
Sync will lock the entire object.  By the way we have implemented lock it is possible we have fundamentally already done that.